File: app/src/main/java/com/fun/hairclipper/UI/HalloweenActivity.java
package com.fun.hairclipper.UI;

import android.annotation.SuppressLint;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.cardview.widget.CardView;

import com.airbnb.lottie.LottieAnimationView;
import com.fun.hairclipper.R;
import com.fun.hairclipper.tools.Tool;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.interstitial.InterstitialAd;

public class HalloweenActivity extends BaseClass {
    MediaPlayer player, loopPlayer;
    private SeekBar volumeSeekbar = null;
    private AudioManager audioManager = null;
    String soundSelect = "bike";
    Boolean loopStatus = false;
    ImageView imageView, loopImageView;
    private FrameLayout frameLayout;
    private InterstitialAd mInterstitialAd;

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_halloween);
        setVolumeControlStream(AudioManager.STREAM_MUSIC);
        imageView = findViewById(R.id.imagehiorn);
        loopImageView = findViewById(R.id.loopImage);
        frameLayout = findViewById(R.id.adaptiveHalloween);
        if (!getPaymentSubscription().isPurchased() && Tool.isNetworkAvailable(this)) {
            loadBanner();
        } else {
            frameLayout.setVisibility(View.GONE);
        }
        initControls();
        ImageView home = findViewById(R.id.home_btn);
        home.setOnClickListener(view -> {
            finish();
        });
        LottieAnimationView splashscreeny;
        splashscreeny = findViewById(R.id.lott);
        splashscreeny.animate();
        Animation shake;
        shake = AnimationUtils.loadAnimation(getApplicationContext(), R.anim.shake);
        Button loopButton = findViewById(R.id.loopButton);
        loopButton.setText(R.string.start_loop);
        loopButton.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_loop, 0, 0, 0);
        loopButton.setOnClickListener(v -> {

            if (!loopStatus) {
                loopImageView.setVisibility(View.VISIBLE);
                loopButton.setText(R.string.stop_loop);
                loopButton.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_stop, 0, 0, 0);
                loopStatus = true;
                splashscreeny.setVisibility(View.VISIBLE);
                if (soundSelect.equals("bike")) {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.shiver);
                } else if (soundSelect.equals("police")) {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.creepy);
                } else if (soundSelect.equals("ambulance")) {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.horror);
                } else {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.scary);
                }

                player.start();
                player.setLooping(true);
                imageView.startAnimation(shake);
                splashscreeny.playAnimation();
            } else {
                loopImageView.setVisibility(View.GONE);
                loopButton.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_loop, 0, 0, 0);
                loopButton.setText(R.string.start_loop);
                loopStatus = false;
                player.stop();
                player.release();
                imageView.clearAnimation();
                splashscreeny.pauseAnimation();
                splashscreeny.setVisibility(View.GONE);

            }


//            if (!loopPlayer.isPlaying()){
//                loopPlayer.start();
//                    loopPlayer.setLooping(true);
//                    loopButton.setText(R.string.stop_loop);
//                loopButton.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_stop, 0, 0, 0);
//
//            }else {
//                loopPlayer.pause();
//                //loopPlayer.release();
//                loopButton.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_loop, 0, 0, 0);
//                loopButton.setText(R.string.start_loop);
//
//            }
        });

        loopImageView.setOnClickListener(v -> Toast.makeText(HalloweenActivity.this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show());
        imageView.setOnTouchListener((v, event) -> {

            if (event.getAction() == MotionEvent.ACTION_DOWN) {
                splashscreeny.setVisibility(View.VISIBLE);
                if (soundSelect.equals("bike")) {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.shiver);
                } else if (soundSelect.equals("police")) {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.creepy);
                } else if (soundSelect.equals("ambulance")) {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.horror);
                } else {
                    player = MediaPlayer.create(HalloweenActivity.this, R.raw.scary);
                }

                player.start();
                player.setLooping(true);
                imageView.startAnimation(shake);
                splashscreeny.playAnimation();

            } else if (event.getAction() == MotionEvent.ACTION_UP) {
                player.stop();
                player.release();
                imageView.clearAnimation();
                splashscreeny.pauseAnimation();
                splashscreeny.setVisibility(View.GONE);


            }
            return true;

        });

    }

    @Override
    public void onBackPressed() {
        if (loopStatus) {
            cleanUpMediaPlayer();
        }
        super.onBackPressed();
    }

    private void initControls() {
        loopPlayer = MediaPlayer.create(HalloweenActivity.this, R.raw.shiver);

        CardView card1 = findViewById(R.id.sound1);
        CardView card2 = findViewById(R.id.sound2);
        CardView card3 = findViewById(R.id.sound3);
        CardView card4 = findViewById(R.id.sound4);

        TextView text1 = findViewById(R.id.text_sound1);
        TextView text2 = findViewById(R.id.text_sound2);
        TextView text3 = findViewById(R.id.text_sound3);
        TextView text4 = findViewById(R.id.text_sound4);

        text1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, R.drawable.ic_down);
        imageView.setBackgroundResource(R.drawable.halloween1);
        soundSelect = "bike";
        card1.setOnClickListener(v -> {
            if (!loopStatus) {

                imageView.setBackgroundResource(R.drawable.halloween1);
                loopPlayer = MediaPlayer.create(HalloweenActivity.this, R.raw.shiver);
                text1.setText("");
                text2.setText(R.string.creeps);
                text3.setText(R.string.horror);
                text4.setText(R.string.scary);
                text1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, R.drawable.ic_down);
                text2.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text3.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text4.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                soundSelect = "bike";

            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show();
            }

        });
        card2.setOnClickListener(v -> {
            if (!loopStatus) {

                imageView.setBackgroundResource(R.drawable.halloween2);
                loopPlayer = MediaPlayer.create(HalloweenActivity.this, R.raw.creepy);
                text1.setText(R.string.shiver);
                text2.setText("");
                text3.setText(R.string.horror);
                text4.setText(R.string.scary);
                text1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text2.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, R.drawable.ic_down);
                text3.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text4.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                soundSelect = "police";
            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show();
            }


        });
        card3.setOnClickListener(v -> {
            if (!loopStatus) {

                imageView.setBackgroundResource(R.drawable.halloween3);
                loopPlayer = MediaPlayer.create(HalloweenActivity.this, R.raw.horror);
                text1.setText(R.string.shiver);
                text2.setText(R.string.creeps);
                text3.setText("");
                text4.setText(R.string.scary);
                text1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text2.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text3.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, R.drawable.ic_down);
                text4.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                soundSelect = "ambulance";
            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show();

            }

        });
        card4.setOnClickListener(v -> {
            if (!loopStatus) {

                imageView.setBackgroundResource(R.drawable.halloween4);
                loopPlayer = MediaPlayer.create(HalloweenActivity.this, R.raw.scary);
                text1.setText(R.string.shiver);
                text2.setText(R.string.creeps);
                text3.setText(R.string.horror);
                text4.setText("");
                text1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text2.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text3.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                text4.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, R.drawable.ic_down);
                soundSelect = "truck";
            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show();

            }

        });

        ImageView low = findViewById(R.id.volumeLow);
        low.setOnClickListener(v -> {
            int index = volumeSeekbar.getProgress();
            volumeSeekbar.setProgress(index - 1);
        });
        ImageView high = findViewById(R.id.volumeUp);
        high.setOnClickListener(v -> {
            int index = volumeSeekbar.getProgress();
            volumeSeekbar.setProgress(index + 1);
        });
        try {
            volumeSeekbar = findViewById(R.id.seekbar1);
            audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);

            volumeSeekbar.setMax(audioManager
                    .getStreamMaxVolume(AudioManager.STREAM_MUSIC));
            volumeSeekbar.setProgress(audioManager
                    .getStreamVolume(AudioManager.STREAM_MUSIC));
            volumeSeekbar.setMax(20);
            volumeSeekbar.setProgress(15);
            int pro = volumeSeekbar.getProgress();

            audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, pro, 0);

            volumeSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

                @Override
                public void onStopTrackingTouch(SeekBar arg0) {
                }

                @Override
                public void onStartTrackingTouch(SeekBar arg0) {
                }

                @Override
                public void onProgressChanged(SeekBar arg0, int progress, boolean arg2) {
                    audioManager.setStreamVolume(AudioManager.STREAM_MUSIC,
                            progress, 0);
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    private void loadBanner() {
        AdView adView = new AdView(this);
        adView.setAdUnitId(getString(R.string.BannerAd));
        frameLayout.addView(adView);

        AdRequest adRequest = new AdRequest.Builder().build();
        AdSize adSize = getAdSize();
        adView.setAdSize(adSize);
        adView.loadAd(adRequest);
    }

    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
            // mediaVlmSeekBar = (SeekBar) findViewById(R.id.seekBar1);
            int index = volumeSeekbar.getProgress();
            volumeSeekbar.setProgress(index + 1);
            return true;
        } else if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
            int index = volumeSeekbar.getProgress();
            volumeSeekbar.setProgress(index - 1);
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    public void cleanUpMediaPlayer() {
        if (player != null) {
            if (player.isPlaying()) {
                player.stop();
            }
            player.release();
        }
    }
}

File: app/src/main/java/com/fun/hairclipper/UI/HalloweenActivity.kt
package com.`fun`.hairclipper.UI

import android.annotation.SuppressLint
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.Bundle
import android.view.KeyEvent
import android.view.MotionEvent
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.SeekBar
import android.widget.Toast
import com.`fun`.hairclipper.R
import com.`fun`.hairclipper.admobHelper.AdConstants
import com.`fun`.hairclipper.admobHelper.BannerAd
import com.`fun`.hairclipper.admobHelper.FullScreenAdListener
import com.`fun`.hairclipper.admobHelper.MyApplication
import com.`fun`.hairclipper.admobHelper.RemoteConfig
import com.`fun`.hairclipper.admobHelper.internetConnection
import com.`fun`.hairclipper.databinding.ActivityHalloweenBinding

class HalloweenActivity : BaseClass() {
    private val binding by lazy { ActivityHalloweenBinding.inflate(layoutInflater) }
    var player: MediaPlayer? = null
    var loopPlayer: MediaPlayer? = null
    private var audioManager: AudioManager? = null
    var soundSelect: String = "bike"
    var loopStatus: Boolean = false

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        volumeControlStream = AudioManager.STREAM_MUSIC
        loadBanner()
        initControls()
        binding.homeBtn.setOnClickListener { handleBackPressed() }
        binding.lott.animate()
        val shake: Animation = AnimationUtils.loadAnimation(applicationContext, R.anim.shake)

        binding.loopButton.apply {
            setText(R.string.start_loop)
            setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_loop, 0, 0, 0)
            setOnClickListener {
                if (!loopStatus) {
                    binding.loopImage.visibility = View.VISIBLE
                    setText(R.string.stop_loop)
                    setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_stop, 0, 0, 0)
                    loopStatus = true
                    binding.lott.visibility = View.VISIBLE

                    player = when (soundSelect) {
                        "bike" -> MediaPlayer.create(this@HalloweenActivity, R.raw.shiver)
                        "police" -> MediaPlayer.create(this@HalloweenActivity, R.raw.creepy)
                        "ambulance" -> MediaPlayer.create(this@HalloweenActivity, R.raw.horror)
                        else -> MediaPlayer.create(this@HalloweenActivity, R.raw.scary)
                    }

                    player!!.start()
                    player!!.isLooping = true
                    binding.imagehiorn.startAnimation(shake)
                    binding.lott.playAnimation()
                } else {
                    MyApplication.showInterstitialAdHalloween(this@HalloweenActivity,object :
                        FullScreenAdListener(){
                        override fun gotoNext() {
                            super.gotoNext()
                            binding.loopImage.visibility = View.GONE
                            setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_loop, 0, 0, 0)
                            setText(R.string.start_loop)
                            loopStatus = false
                            player!!.stop()
                            player!!.release()
                            binding.imagehiorn.clearAnimation()
                            binding.lott.pauseAnimation()
                            binding.lott.visibility = View.GONE
                        }
                    },"btn_stop_loop")

                }
            }
        }

        binding.loopImage.setOnClickListener {
            Toast.makeText(this@HalloweenActivity, R.string.stop_loop_first, Toast.LENGTH_SHORT)
                .show()
        }

        binding.imagehiorn.setOnTouchListener { _, event: MotionEvent? ->
            if (loopStatus) return@setOnTouchListener true

            when (event?.action) {
                MotionEvent.ACTION_DOWN -> {
                    binding.lott.visibility = View.VISIBLE
                    player = when (soundSelect) {
                        "bike" -> MediaPlayer.create(this@HalloweenActivity, R.raw.shiver)
                        "police" -> MediaPlayer.create(this@HalloweenActivity, R.raw.creepy)
                        "ambulance" -> MediaPlayer.create(this@HalloweenActivity, R.raw.horror)
                        else -> MediaPlayer.create(this@HalloweenActivity, R.raw.scary)
                    }
                    player!!.start()
                    player!!.isLooping = true
                    binding.imagehiorn.startAnimation(shake)
                    binding.lott.playAnimation()
                }

                MotionEvent.ACTION_UP -> {
                    player!!.stop()
                    player!!.release()
                    binding.imagehiorn.clearAnimation()
                    binding.lott.pauseAnimation()
                    binding.lott.visibility = View.GONE
                }
            }
            true
        }
    }

    override fun handleBackPressed() {
        MyApplication.showInterstitialAdHalloween(this,object : FullScreenAdListener(){
            override fun gotoNext() {
                super.gotoNext()
                if (loopStatus) {
                    cleanUpMediaPlayer()
                }
                finish()
            }
        },"btn_back_halloween")
    }

    private fun initControls() {
        loopPlayer = MediaPlayer.create(this@HalloweenActivity, R.raw.shiver)
        binding.textSound1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, R.drawable.ic_down)
        binding.imagehiorn.setBackgroundResource(R.drawable.halloween1)
        soundSelect = "bike"

        binding.sound1.setOnClickListener {
            if (!loopStatus) {
                binding.imagehiorn.setBackgroundResource(R.drawable.halloween1)
                loopPlayer = MediaPlayer.create(this@HalloweenActivity, R.raw.shiver)
                binding.textSound1.text = ""
                binding.textSound2.setText(R.string.creeps)
                binding.textSound3.setText(R.string.horror)
                binding.textSound4.setText(R.string.scary)
                binding.textSound1.setCompoundDrawablesWithIntrinsicBounds(
                    0,
                    0,
                    0,
                    R.drawable.ic_down
                )
                binding.textSound2.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound3.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound4.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                soundSelect = "bike"
            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show()
            }
        }

        binding.sound2.setOnClickListener {
            if (!loopStatus) {
                binding.imagehiorn.setBackgroundResource(R.drawable.halloween2)
                loopPlayer = MediaPlayer.create(this@HalloweenActivity, R.raw.creepy)
                binding.textSound1.setText(R.string.shiver)
                binding.textSound2.text = ""
                binding.textSound3.setText(R.string.horror)
                binding.textSound4.setText(R.string.scary)
                binding.textSound1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound2.setCompoundDrawablesWithIntrinsicBounds(
                    0,
                    0,
                    0,
                    R.drawable.ic_down
                )
                binding.textSound3.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound4.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                soundSelect = "police"
            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show()
            }
        }

        binding.sound3.setOnClickListener {
            if (!loopStatus) {
                binding.imagehiorn.setBackgroundResource(R.drawable.halloween3)
                loopPlayer = MediaPlayer.create(this@HalloweenActivity, R.raw.horror)
                binding.textSound1.setText(R.string.shiver)
                binding.textSound2.setText(R.string.creeps)
                binding.textSound3.text = ""
                binding.textSound4.setText(R.string.scary)
                binding.textSound1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound2.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound3.setCompoundDrawablesWithIntrinsicBounds(
                    0,
                    0,
                    0,
                    R.drawable.ic_down
                )
                binding.textSound4.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                soundSelect = "ambulance"
            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show()
            }
        }

        binding.sound4.setOnClickListener {
            if (!loopStatus) {
                binding.imagehiorn.setBackgroundResource(R.drawable.halloween4)
                loopPlayer = MediaPlayer.create(this@HalloweenActivity, R.raw.scary)
                binding.textSound1.setText(R.string.shiver)
                binding.textSound2.setText(R.string.creeps)
                binding.textSound3.setText(R.string.horror)
                binding.textSound4.text = ""
                binding.textSound1.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound2.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound3.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                binding.textSound4.setCompoundDrawablesWithIntrinsicBounds(
                    0,
                    0,
                    0,
                    R.drawable.ic_down
                )
                soundSelect = "truck"
            } else {
                Toast.makeText(this, R.string.stop_loop_first, Toast.LENGTH_SHORT).show()
            }
        }

        binding.volumeLow.setOnClickListener { binding.seekbar1.progress -= 1 }
        binding.volumeUp.setOnClickListener { binding.seekbar1.progress += 1 }

        try {
            audioManager = getSystemService(AUDIO_SERVICE) as? AudioManager
            audioManager?.let { am ->
                binding.seekbar1.max = 20
                binding.seekbar1.progress = 15
                val pro = binding.seekbar1.progress
                am.setStreamVolume(AudioManager.STREAM_MUSIC, pro, 0)

                binding.seekbar1.setOnSeekBarChangeListener(object :
                    SeekBar.OnSeekBarChangeListener {
                    override fun onStopTrackingTouch(arg0: SeekBar?) {}
                    override fun onStartTrackingTouch(arg0: SeekBar?) {}
                    override fun onProgressChanged(arg0: SeekBar?, progress: Int, arg2: Boolean) {
                        am.setStreamVolume(AudioManager.STREAM_MUSIC, progress, 0)
                    }
                })
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun loadBanner() {
        val adId = if (AdConstants.TEST_ADS) {
            getString(R.string.BannerAd)
        } else {
            RemoteConfig.getString(RemoteConfig.HALLOWEEN_BANNER_AD_ID)
        }
        val makeCollapsible = RemoteConfig.getBoolean(RemoteConfig.HALLOWEEN_BANNER_MAKE_COLLAPSIBLE)
        val enable = RemoteConfig.getBoolean(RemoteConfig.ENABLE_HALLOWEEN_BANNER_AD)

        if (paymentSubscription.isPurchased.not() && internetConnection(this) && enable) {
            BannerAd.load(binding.adaptiveHalloween, adId.trim(), makeCollapsible)
        } else {
            binding.adaptiveHalloween.visibility = View.GONE
        }
    }


    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        when (keyCode) {
            KeyEvent.KEYCODE_VOLUME_UP -> {
                binding.seekbar1.progress += 1
                return true
            }

            KeyEvent.KEYCODE_VOLUME_DOWN -> {
                binding.seekbar1.progress -= 1
                return true
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    fun cleanUpMediaPlayer() {
        if (player != null) {
            if (player!!.isPlaying) {
                player!!.stop()
            }
            player!!.release()
            player = null
        }
    }
}